birdset/modules/hybrid_module.py

```diff
@@ -0,0 +1,117 @@
+                torch.softmax,
+                dim=1
+            ),
+            loss: _Loss = CrossEntropyLoss(),
+            optimizer: partial[Type[Optimizer]] = partial(
+                AdamW,
+                lr=1e-2,
+                weight_decay=5e-4,
+            ),
+            lr_scheduler: Optional[LRSchedulerConfig] = LRSchedulerConfig(),
+            metrics: MulticlassMetricsConfig | MultilabelMetricsConfig = MulticlassMetricsConfig(),
+            logging_params: LoggingParamsConfig = LoggingParamsConfig(),
+            num_epochs: int = 50,
+            len_trainset: int = 13878, # set to property from datamodule
+            batch_size: int = 32,
+            task: Literal['multiclass', 'multilabel'] = "multiclass",
+            num_gpus: int = 1,
+            pretrain_info = None,
+            embedding_model: EmbeddingModuleConfig = EmbeddingModuleConfig(), # Model for extracting the embeddings
+            ft_lr: int = 1e-2,
+            ft_at_epoch: int = 10,
+            ):
+        super().__init__(
+            network = network,
+            output_activation=output_activation,
+            loss=loss,
+            optimizer=optimizer,
+            lr_scheduler=lr_scheduler,
+            metrics=metrics,
+            logging_params=logging_params,
+            num_epochs=num_epochs,
+            len_trainset=len_trainset,
+            batch_size=batch_size,
+            task=task,
+            num_gpus=num_gpus,
+            pretrain_info = pretrain_info,
+            embedding_model=embedding_model
+        )
+        self.embedding_model.train()
+        self.ft_lr = ft_lr
+        self.ft_at_epoch = ft_at_epoch
+        # Freeze the embedding model at the start
+        self.freeze_embedding_model()
+        
+    def configure_optimizers(self):
+        self.optimizer = self.optimizer(self.model.parameters())
+        return {"optimizer": self.optimizer}
+        #optimizer = self.optimizer([{'name':'classifier','params': list(self.model.parameters())+list(self.embedding_model.parameters()), 'lr': 1e-2}])#, {'name':'embedding','params': self.embedding_model.parameters(), 'lr': 0.0}])
+
+        # Custom scheduler: starts with warmup, then reduces to 0, and gradually increases after unfreezing
+        '''def lr_lambda(epoch):
+            if epoch < self.ft_max_epochs:
+                # Linearly increase the learning rate during warmup
+                return 1.0
+            elif epoch == self.ft_max_epochs:
+                # Reduce LR to zero after warmup
+                return 0.0
+            else:
+                # Gradually increase LR from 0 to max_lr after unfreezing
+                total_epochs = self.trainer.max_epochs
+                progress = (epoch - self.ft_max_epochs+1) / (total_epochs - self.ft_max_epochs+1)
+                return progress * (self.ft_lr / 1e-2)
+        
+        scheduler = LambdaLR(optimizer, lr_lambda=lr_lambda)
+        return [optimizer], [scheduler]'''
```

Comment:
remove all commented out code, if you think its worth saving, put it into another function and add a config parameter to be able to use it.